CommunityToolkit.Mvvm/Messaging/Internals/System/Collections.Generic/Dictionary2.cs

```diff
@@ -0,0 +1,478 @@
+        /// </summary>
+        /// <param name="dictionary">The input dictionary to enumerate.</param>
+        internal Enumerator(Dictionary2<TKey, TValue> dictionary)
+        {
+            this.entries = dictionary.entries;
+            this.index = 0;
+            this.count = dictionary.count;
+        }
+
+        /// <inheritdoc cref="IEnumerator.MoveNext"/>
+        public bool MoveNext()
+        {
+            while ((uint)this.index < (uint)this.count)
+            {
+                // We need to preemptively increment the current index so that we still correctly keep track
+                // of the current position in the dictionary even if the users doesn't access any of the
+                // available properties in the enumerator. As this is a possibility, we can't rely on one of
+                // them to increment the index before MoveNext is invoked again. We ditch the standard enumerator
+                // API surface here to expose the Key/Value properties directly and minimize the memory copies.
+                // For the same reason, we also removed the KeyValuePair<TKey, TValue> field here, and instead
+                // rely on the properties lazily accessing the target instances directly from the current entry
+                // pointed at by the index property (adjusted backwards to account for the increment here).
+                if (this.entries![this.index++].Next >= -1)
+                {
+                    return true;
+                }
+            }
+
+            this.index = this.count + 1;
+
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the current key.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public TKey GetKey()
+        {
+            return this.entries[this.index - 1].Key;
+        }
+
+        /// <summary>
+        /// Gets the current value.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public TValue GetValue()
+        {
+            return this.entries[this.index - 1].Value!;
+        }
+    }
+
+    /// <summary>
+    /// Gets the value for the specified key, or.
+    /// </summary>
+    /// <param name="key">Key to look for.</param>
+    /// <returns>Reference to the existing value.</returns>
+    private ref TValue FindValue(TKey key)
+    {
+        ref Entry entry = ref Unsafe.NullRef<Entry>();
+        uint hashCode = (uint)key.GetHashCode();
+        int i = GetBucket(hashCode);
+        Entry[] entries = this.entries;
+
+        i--;
+        do
+        {
+            if ((uint)i >= (uint)entries.Length)
+            {
+                goto ReturnNotFound;
+            }
+
+            entry = ref entries[i];
+
+            if (entry.HashCode == hashCode && entry.Key.Equals(key))
+            {
+                goto ReturnFound;
+            }
+
+            i = entry.Next;
+        }
+        while (true);
```

Comment:
It would be nice to see a little more testing around just this do-while block, especially a terminating condition, to make sure we don't get stuck in here.

Reply:
That's implicitly tested, but still good to point out, so good catch! 👍

This type is a custom fork I did of the `Dictionary<TKey, TValue>` type from the BCL, with some specific optimizations I added to better support `WeakReferenceMessenger`. In the original type, this loop here uses the number of collisions as a terminating condition: [link here](https://github.com/dotnet/runtime/blob/12a8819eee9865eb38bca6c05fdece1053102854/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L396-L400). This is done to guard against concurrent modifications, which are not supported (the type is not thread-safe). In fact, you can see that right after that loop ends, the code just jumps to a path that throws an exception, with the reasoning being that a concurrent update was detected. In our case, I'm relying on the fact that the messenger type already uses a lock for all register/unregister/send operations, and therefore all accesses to this dictionary are already guaranteed to be thread-safe, meaning this faulting condition will never happen, and the loop will always terminate (either by finding a value, or by reaching the end of a linked chain and not finding any value). This is implicitly tested as part of our suite of concurrrent tests, here:

https://github.com/CommunityToolkit/dotnet/blob/8eaea35c2a08ff11a29792dd2abecb26ce08733e/tests/CommunityToolkit.Mvvm.UnitTests/Test_Messenger.cs#L715-L723

And the other 3 following unit tests there.

Does this make sense? 🙂

Reply:
I'm sure there was implicit testing going on, if you're designing your own dictionary class (RIP, `DictionarySlim`).

> and the loop will always terminate (either by ..., or by reaching the end of a linked chain and not finding any value). 

Not saying we need to safeguard the code to prevent any and all misuse, but if a developer wound up emulating a circular linked list (i.e., last dictionary item Next property points to the first dictionary item), couldn't this do-while loop ping-pong back and forth with a non-existent key value, never finding a match, endlessly, until the End of Days?

Reply:
RIP `DictionarySlim` indeed 😄
On the bright side though, `Dictionary2` has pretty much the same performance in the general case, with the benefit of being much faster when developers are using token types that are more expensive to hash (eg. `string`-s).